drivers/timer/ite_it8xxx2_timer.c

```diff
@@ -246,21 +251,26 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 		 */
 		k_spin_unlock(&lock, key);
 		return;
-	} else if (ticks <= 1) {
-		/*
-		 * Ticks <= 1 means the kernel wants the tick announced
-		 * as soon as possible, ideally no more than one system tick
-		 * in the future. So set event timer count to 1 system tick or
-		 * at least 1 hw count.
-		 */
-		hw_cnt = MAX((1 * HW_CNT_PER_SYS_TICK), 1);
 	} else {
+		uint32_t next_cycs;
+		uint32_t now;
+		uint32_t dcycles;
+
 		/*
-		 * Set event timer count to EVENT_TIMER_MAX_CNT, after
-		 * interrupt fired the remaining time will be set again
-		 * by sys_clock_announce().
+		 * If ticks <= 1 means the kernel wants the tick announced
+		 * as soon as possible, ideally no more than one system tick
+		 * in the future. So set event timer count to 1 HW tick.
 		 */
-		hw_cnt = MIN((ticks * HW_CNT_PER_SYS_TICK), EVENT_TIMER_MAX_CNT);
+		ticks = CLAMP(ticks, 1, (int32_t)EVEN_TIMER_MAX_CNT_SYS_TICK);
+
+		next_cycs = (last_ticks + last_elapsed + ticks) * HW_CNT_PER_SYS_TICK;
+		now = ~(IT8XXX2_EXT_CNTOX(FREE_RUN_TIMER));
+		if (unlikely(next_cycs <= now)) {
+			hw_cnt = 1;
```

Comment:
Nitpick: expressing this as a CLAMP() instead of the likely-tagged branch is probably going to generate better code.  Even more so if you can roll the clamp into the ticks computation above.

Just in general the unlikely() construction tends to hurt and not help (both code and readability) if the blocks are tiny one-liners.  Let the branch predictor do its job.

Finally: does a clamp of 1 actually work?  Most timer hardware is based on an equality test in the counter and will drop interrupts if you try to set a deadline too close in the future, and in any case "1" is going to have aliasing bugs where it really lands in the register at a time when it should have been "0".  Some hardware does a proper comparator though, so maybe not necessary.